torchtune/modules/transforms/_transforms.py

```diff
@@ -0,0 +1,164 @@
+            )
+        ]
+        # Last image will attend to subsequent text until end
+        vision_masks.append([vision_token_locations[-1], end])
+
+        # If there are consecutive vision tokens, they should all attend to the
+        # same subsequent text
+        last_mask_end = vision_masks[-1][1]
+        for vision_mask in vision_masks[::-1]:
+            if vision_mask[0] == vision_mask[1] - 1:
+                vision_mask[1] = last_mask_end
+            last_mask_end = vision_mask[1]
+        return vision_masks
+
+    def __call__(self, *, tokens: List[int], images: List[torch.Tensor], **kwargs):
+        """
+        Generates the vision cross-attention mask for the given sample based on
+        the image token locations interleaved in the text sequence.
+
+        Args:
+            tokens (List[int]): List of token IDs in the text sequence. Number of
+                image token IDs in the sequence must match the number of images.
+            images (List[torch.Tensor]): List of image Tensors post-tiling of shape
+                (n_tiles, c, h, w) each.
+            **kwargs (Dict[str, Any]): all other keys within the sample that will
+                not be altered by this transform.
+
+        Returns:
+            Dict[str, Any]: updated sample with the following keys:
+                - encoder_mask (List[torch.Tensor]): masks of shape (text_seq_len, image_seq_len)
+                - tokens (List[int]): original tokens
+                - images (List[torch.Tensor]): original images
+        """
+
+        # One sample can have multiple images - verify the number of image tokens
+        # is the same
+        n_img = len(images)
+        intervals = self._get_image_attention_intervals(tokens)
+        assert len(intervals) == n_img
+
+        # Create mask for each individual image based on its number of tokens,
+        # which can vary based on number of tiles. The masks are padded and concatenated
+        # together in the batch collator
+        text_seq_len = len(tokens)
+        masks = []
+        for image_num, interval in enumerate(intervals):
+            # Identify what part of text sequence should be attended
+            start, end = interval
+            # Compute this image's number of tokens based on num tiles, patches per tile
+            n_tiles = images[image_num].shape[0]
+            image_seq_len = n_tiles * (self.patches_per_tile + 1)  # +1 for CLS token
+            # Mask will be block of 1s at the corresponding interval in the text.
+            # It is not a causal block because all the image tokens correspond
+            # to a single image, so text tokens attend to all the image's tokens
+            mask = torch.zeros(text_seq_len, image_seq_len, dtype=torch.bool)
+            mask[start:end, :] = True
+            masks.append(mask)
+
+        kwargs.update({"encoder_mask": masks, "tokens": tokens, "images": images})
```

Comment:
Maybe it's just me but I find this whole kwargs pattern kinda unintuitive. Like really we are covertly using this transform to also add its inputs into the dictionary (does that mean that it doesn't already contain them?). Maybe I need to see the callsite, but I don't understand why we can't just have e.g. 

```
sample: Dict[str, Any]
mask_transform = VisionCrossAttentionMask(...)
sample["encoder_mask"] = mask_transform(sample["tokens"], sample["images"])
``` 

Alternatively if you want to go more of the compose route:

```
class VisionCrossAttentionMask:
	def __call__(sample: Dict[str, Any]) -> Dict[str, Any]:
		# just key into sample directly in the transform
		sample["encoder_mask"] = construct_mask(sample["tokens"], sample["images"])
		return sample

mask_transform = VisionCrossAttentionMask(...)
sample = mask_transform(sample)
```

But rn we are kinda halfway in between the two and it feels clunky to me.

Reply:
cc @pbontrager  who'sthought about this a lot more

We were between two approaches, the kwargs approach here and the sample approach you described where you key into the fields you need for the transform. One issue with that was this implicit assumption that token ids are always under "tokens", image tensors are always under "images", etc., so we thought we may have to provide these keys as attributes to the transform class. Then the key management gets a bit messy to handle and error prone.

Although, thinking about this more, the collator will also implicitly assume the keys "tokens", "labels", "images" etc will be present. If we're consistent with these across the library, and the fact that the number of these keys in the sample dict will remain quite small ~5-10 max, it might be fine to keep these assumptions, or do something similar to checkpointing with model key, optim key, etc as constants.

I don't have a strong opinion either way, although it would be nice to keep a forward signature of just sample.

Reply:
Yeah imo you're gonna need the keys somewhere anyways, better to just be explicit about it than do this weird back and forth between dict values and standalone arguments.